spec/src/main/asciidoc/repository.asciidoc

```diff
@@ -833,17 +833,18 @@ where:
 - `R m(P1 p1, P2 p2, ...)` is a legal repository method signature according to the previous sections of this chapter, or, in the special case that `R` is the type `java.lang.Void`, `void m(P1 p1, P2 p2, ...)` is a legal repository method signature according to the previous sections of this chapter, and
 - `F` is a parameterized type representing a value which might not yet have been computed and which is supported by the Jakarta Data provider.
 
-NOTE: Every Jakarta Data provider is encouraged, but not required, to support asynchronous repository methods returning `java.util.concurrent.CompletionStage`.
+NOTE: Repository methods that are annotated with the `@Asynchronous` annotation from the Jakarta Concurrency specification must be able to return the types `java.util.concurrent.CompletionStage` and `java.util.concurrent.CompletableFuture`, in which case the Jakarta Data provider returns an already completed `CompletionStage` or `CompletableFuture` such that the Jakarta Concurrency provider is able to control the asynchronous behavior. Every Jakarta Data provider is otherwise encouraged, but not required, to support asynchronous repository methods returning `java.util.concurrent.CompletionStage`.
```

Comment:
```suggestion
NOTE: Repository methods annotated with the `@Asynchronous` annotation from the Jakarta Concurrency specification must be able to return the types `java.util.concurrent.CompletionStage` and `java.util.concurrent.CompletableFuture`, in which case the Jakarta Data provider returns an already completed `CompletionStage` or `CompletableFuture` such that the Jakarta Concurrency provider is able to control the asynchronous behavior. Every Jakarta Data provider is otherwise encouraged, but not required, to support asynchronous repository methods returning `java.util.concurrent.CompletionStage`.
```

Reply:
I'm not sure we should allow `CompletableFuture` here. I think `CompletionStage` is enough.

Reply:
The Jakarta Concurrency spec explicitly states `CompletableFuture` as one of the 3 return types to use with [@Asynchronous](https://jakarta.ee/specifications/concurrency/3.1/apidocs/jakarta.concurrency/jakarta/enterprise/concurrent/asynchronous), so it seems correct to mention it:

> Methods with the following return types can be annotated to be asynchronous methods:
>
>  * [CompletableFuture](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/util/concurrent/CompletableFuture.html)
>  * [CompletionStage](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/util/concurrent/CompletionStage.html)
>  * void

Reply:
> The Jakarta Concurrency spec explicitly states CompletableFuture as one of the 3 return types

Yes, sure, I saw that, but that doesn't mean that Jakarta Data should require that providers support this.

And, I'm not an expert on this stuff, but it seems to me that it's better to _not_ expose the `complete()` method to repository clients. It's the responsibility of the repository implementation to call this operation, and clients should not be messing with it. 

AFAIK, `CompletionStage` is the right level of abstraction for clients.

Reply:
> it seems to me that it's better to _not_ expose the `complete()` method to repository clients

Yeah, I agree. I was only thinking about it relation to Find operations. But if a user has an asynchronous update/delete/insert, we wouldn't want them to be able to forcibly complete or cancel its CompletableFuture because it would be uncertain to the application whether the change was made or not.  I added commit https://github.com/jakartaee/data/pull/1003/commits/79c9fa5be19bb4717ed28471c949e9436fe7049c to remove CompletableFuture.